Source/FieldSolver/ImplicitSolvers/SemiImplicitEM.cpp

```diff
@@ -0,0 +1,167 @@
+{
+    if (!m_WarpX->Verbose()) { return; }
+    amrex::Print() << std::endl;
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << "----------- SEMI IMPLICIT EM SOLVER PARAMETERS ------------" << std::endl;
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << "max particle iterations:    " << m_max_particle_iterations << std::endl;
+    amrex::Print() << "particle tolerance:         " << m_particle_tolerance << std::endl;
+    if (m_nlsolver_type==NonlinearSolverType::Picard) {
+        amrex::Print() << "Nonlinear solver type:      Picard" << std::endl;
+    }
+    else if (m_nlsolver_type==NonlinearSolverType::Newton) {
+        amrex::Print() << "Nonlinear solver type:      Newton" << std::endl;
+    }
+    m_nlsolver->PrintParams();
+    amrex::Print() << "-----------------------------------------------------------" << std::endl;
+    amrex::Print() << std::endl;
+}
+
+void SemiImplicitEM::OneStep ( amrex::Real  a_time,
+                               amrex::Real  a_dt,
+                               int          a_step )
+{
+    using namespace amrex::literals;
+    amrex::ignore_unused(a_step);
+
+    // Fields have E^{n}, B^{n-1/2]
+    // Particles have p^{n} and x^{n}.
+
+    // Save the values at the start of the time step,
+    m_WarpX->SaveParticlesAtImplicitStepStart ( );
+
+    // Save the fields at the start of the step
+    m_Eold.Copy( m_WarpX->getMultiLevelField(FieldType::Efield_fp) );
+    m_E = m_Eold; // initial guess for E
+
+    // Compute Bfield at time n+1/2
+    m_WarpX->EvolveB(a_dt, DtType::Full);
+    m_WarpX->ApplyMagneticFieldBCs();
+
+    const amrex::Real half_time = a_time + 0.5_rt*a_dt;
+
+    // Solve nonlinear system for E at t_{n+1/2}
+    // Particles will be advanced to t_{n+1/2}
+    m_nlsolver->Solve( m_E, m_Eold, half_time, a_dt );
+
+    // Update WarpX owned Efield_fp and Bfield_fp to t_{n+1/2}
+    UpdateWarpXState( m_E, half_time, a_dt );
+
+    // Update field boundary probes prior to updating fields to t_{n+1}
+    //m_fields->updateBoundaryProbes( a_dt )
+
+    // Advance particles from time n+1/2 to time n+1
+    m_WarpX->FinishImplicitParticleUpdate();
+
+    // Advance E fields from time n+1/2 to time n+1
+    // Eg^{n+1} = 2.0*E_g^{n+1/2} - E_g^n
+    m_E.linComb( 2._rt, m_E, -1._rt, m_Eold );
+    const amrex::Real new_time = a_time + a_dt;
+    UpdateWarpXState( m_E, new_time, a_dt );
+
+}
+
+void SemiImplicitEM::PreRHSOp ( const WarpXSolverVec&  a_E,
+                                amrex::Real            a_time,
+                                amrex::Real            a_dt,
+                                int                    a_nl_iter,
+                                bool                   a_from_jacobian )
+{
+    amrex::ignore_unused(a_E);
+
+    // update derived variable B and then update WarpX owned Efield_fp and Bfield_fp
```

Comment:
This does not update the B field, I think (apart from the mirrors)
```suggestion
    // Update WarpX owned Efield_fp
```